bigtable/client/filters.h

```diff
@@ -410,34 +409,99 @@ class Filter {
    *
    * These filters compose several filters to build complex filter expressions.
    */
-  /// Return a filter that selects between two other filters based on a
-  /// predicate.
-  // TODO(#30) - implement this one.
+  /**
+   * Returns a per-row conditional filter expression.
+   *
+   * For each row the @p predicate filter is evaluated, if it returns any
+   * cells, then the cells returned by @p true_filter are returned, otherwise
+   * the cells from @p false_filter are returned.
+   */
   static Filter Condition(Filter predicate, Filter true_filter,
                           Filter false_filter) {
-    return Filter();
+    Filter tmp;
+    auto& condition = *tmp.filter_.mutable_condition();
+    condition.mutable_predicate_filter()->Swap(&predicate.filter_);
+    condition.mutable_true_filter()->Swap(&true_filter.filter_);
+    condition.mutable_false_filter()->Swap(&false_filter.filter_);
+    return tmp;
   }
 
-  /// Create a chain of filters
-  // TODO(coryan) - document ugly std::enable_if<> hack to ensure they all
-  // are of type Filter.
-  // TODO(#30) - implement this one.
+  /**
+   * Return a chain filter.
+   *
+   * The filter returned by this function acts like a pipeline.  The output
+   * row from each stage is passed on as input for the next stage.
+   *
+   * TODO(#84) - decide what happens when there are no filter arguments.
+   *
+   * @tparam FilterTypes the type of the filter arguments.  They must all be
+   *    convertible to Filter.
+   * @param stages the filter stages.
+   */
   template <typename... FilterTypes>
-  static Filter Chain(FilterTypes&&... a) {
-    return Filter();
+  static Filter Chain(FilterTypes&&... stages) {
+    // This ugly thing provides a better compile-time error message than
+    // just letting the compiler figure things out 3 levels deep
+    // as it recurses on append_types().
+    static_assert(
+        detail::conjunction<std::is_convertible<FilterTypes, Filter>...>::value,
+        "The arguments passed to Chain(...) must be convertible to Filter");
+    Filter tmp;
+    auto& chain = *tmp.filter_.mutable_chain();
+    std::initializer_list<Filter> list{std::forward<FilterTypes>(stages)...};
+    for (Filter const& f : list) {
+      *chain.add_filters() = f.as_proto();
+    }
+    return tmp;
   }
 
-  // TODO(coryan) - same ugly hack documentation needed ...
-  // TODO(#30) - implement this one.
   /**
-   * Return a filter that unions the results of all the other filters.
-   * @tparam FilterTypes
-   * @param a
-   * @return
+   * Return a filter that interleaves the results of many other filters.
+   *
+   * This filter executes each stream in parallel and then merges the results by
+   * interleaving the output from each stream.  The proto
+   * [file](https://github.com/googleapis/googleapis/blob/master/google/bigtable/v2/data.proto)
```

Comment:
Please link the text "proto file" not just "file".

Reply:
Done.